validator/src/consensus/client.test.ts

```diff
@@ -93,41 +114,138 @@ describe("client", () => {
 					verificationShare,
 					peerShares,
 				});
+				return Promise.resolve("0x");
 			},
 		};
 		const clients = validatorAddresses.map(
 			(a) => new FrostClient(a.address, coordinator),
 		);
-		clients.forEach((c) => c.registerParticipants(participants));
+		for (const c of clients) c.registerParticipants(participants);
 		const groupId = keccak256(participantsRoot);
 		console.log(
 			"------------------------ Trigger Keygen Init ------------------------",
 		);
-		clients.forEach((c) => {
+		for (const c of clients) {
 			console.log(`>>>> Keygen init to ${c.validator()} >>>>`);
-			c.handleKeygenInit(groupId, participantsRoot, count, threshold);
-		});
+			await c.handleKeygenInit(groupId, participantsRoot, count, threshold);
+		}
 		console.log(
 			"------------------------ Publish Commitments ------------------------",
 		);
-		clients.forEach((c) =>
-			commitmentEvents.forEach((e) => {
+		for (const c of clients) {
+			for (const e of commitmentEvents) {
 				console.log(
 					`>>>> Keygen commitment from ${e.index} to ${c.participationIndex(e.groupId)} >>>>`,
 				);
 				c.handleKeygenCommitment(e.groupId, e.index, e.commits, e.pok);
-			}),
-		);
+			}
+		}
 		console.log(
 			"------------------------ Publish Shares ------------------------",
 		);
-		clients.forEach((c) =>
-			shareEvents.forEach((e) => {
+		for (const c of clients) {
+			for (const e of shareEvents) {
 				console.log(
 					`>>>> Keygen secrets from ${e.index} to ${c.participationIndex(e.groupId)} >>>>`,
 				);
 				c.handleKeygenSecrets(e.groupId, e.index, e.peerShares);
-			}),
-		);
+			}
+		}
+	});
+
+	it.skip("integration keygen flow", { timeout: 30000 }, async () => {
+		// Make sure to first start the Anvil testnode (run `anvil` in the root)
+		// and run the deployment script: forge script contracts/script/Deploy.s.sol:DeployScript --rpc-url http://127.0.0.1:8545 --private-key 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80 --broadcast
+		// Private key from Anvil testnet
+		const accounts = [
+			privateKeyToAccount(
+				"0x59c6995e998f97a5a0044966f0945389dc9e86dae88c7a8412f4603b6b78690d",
+			),
+			privateKeyToAccount(
+				"0x5de4111afa1a4b94908f83103eb1f1706367c2e68ca870fc3fb9a804cdab365a",
+			),
+			privateKeyToAccount(
+				"0x7c852118294e51e653712a81e05800f419141751be58f605c371e15141b007a6",
+			),
+		];
+		const participants: Participant[] = accounts.map((a, i) => {
+			return { index: BigInt(i + 1), address: a.address };
+		});
+		const coordinatorAddress = "0x601CDd9C9e743AEb40106865775FDB073ee4E889";
```

Comment:
nit: This is pretty fragile (any changes to contracts will require manual intervention for the E2E test. It would be great if we could derive this somehow (off the top of my head - you could load the byte code from the artifact JSON and compute the expected address).